controllers/nnf_node_ec_data_controller.go

```diff
@@ -0,0 +1,205 @@
+/*
+ * Copyright 2022 Hewlett Packard Enterprise Development LP
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package controllers
+
+import (
+	"context"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	nnfec "github.com/NearNodeFlash/nnf-ec/pkg"
+	ec "github.com/NearNodeFlash/nnf-ec/pkg/ec"
+	"github.com/NearNodeFlash/nnf-ec/pkg/persistent"
+	nnfv1alpha1 "github.com/NearNodeFlash/nnf-sos/api/v1alpha1"
+)
+
+const (
+	NnfNodeECDataResourceName = "ec-data"
+)
+
+// NnfNodeECDataReconciler reconciles a NnfNodeECData object
+type NnfNodeECDataReconciler struct {
+	client.Client
+	Scheme  *runtime.Scheme
+	Options *nnfec.Options
+	types.NamespacedName
+}
+
+// Start implements manager.Runnable
+func (r *NnfNodeECDataReconciler) Start(ctx context.Context) error {
+
+	testing := r.NamespacedName.String() == string(types.Separator)
+
+	// During testing, the reconciler is started before kubeapi-server runs, so any calls will fail with
+	// 'connection refused'. The test code instead bootstraps this resource manually
+	if !testing {
+
+		// Create the resource if necessary
+		data := nnfv1alpha1.NnfNodeECData{}
+		if err := r.Get(ctx, r.NamespacedName, &data); err != nil {
+
+			if !errors.IsNotFound(err) {
+				return err
+			}
+
+			data := nnfv1alpha1.NnfNodeECData{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      r.Name,
+					Namespace: r.Namespace,
+				},
+			}
+
+			if err := r.Create(ctx, &data); err != nil {
+				return err
+			}
+		}
+	}
+
+	// This resource acts as the storage provider for the NNF Element Controller
+	persistent.StorageProvider = r
```

Comment:
Do you need the whole reconciler here, or just the client?

Reply:
client & namespaced name